File: tests/brusho-nft-manager/initialize_maker.ts
import * as anchor from "@coral-xyz/anchor";
import { web3 } from "@coral-xyz/anchor";

import { BRUSHO_NFT_MANAGER_PROGRAM, createRealm, getMint, getTokenAccount, GOV_PROGRAM_ID, newSigner, TOKEN_METADATA_PROGRAM_ID } from "../helper";
import { ASSOCIATED_TOKEN_PROGRAM_ID, getAssociatedTokenAddress, TOKEN_PROGRAM_ID } from "@solana/spl-token";
import { assert } from "chai";

export async function createMaker(
    payer: web3.Keypair,
    issuingAuthority: web3.PublicKey,
    updateAuthority: web3.PublicKey,
    name: string,
    metadataUrl: string
): Promise<{
    realm: web3.PublicKey,
    mint: web3.PublicKey,
    councilMint: web3.PublicKey,
    realmAuthority: web3.Keypair,
    maker: web3.PublicKey,
    makerBump: number,
    collection: web3.PublicKey,
    collectionBump: number,
    metadata: web3.PublicKey,
    edition: web3.PublicKey,
    tokenAccount: web3.PublicKey
}> {
    const realmAuthority = await newSigner();
    const [mint, councilMint, realm] = await createRealm(realmAuthority);
    const makerSeeds = [Buffer.from("maker"), realm.toBytes(), Buffer.from(name)];
    const [maker, makerBump] = anchor.web3.PublicKey.findProgramAddressSync(makerSeeds, BRUSHO_NFT_MANAGER_PROGRAM.programId);
    const collectionSeeds = [Buffer.from("collection"), maker.toBytes()];
    const [collection, collectionBump] = anchor.web3.PublicKey.findProgramAddressSync(collectionSeeds, BRUSHO_NFT_MANAGER_PROGRAM.programId);
    const metadataSeeds = [Buffer.from("metadata"), TOKEN_METADATA_PROGRAM_ID.toBytes(), collection.toBytes()];
    const [metadata,] = anchor.web3.PublicKey.findProgramAddressSync(metadataSeeds, TOKEN_METADATA_PROGRAM_ID);
    const editionSeeds = [Buffer.from("metadata"), TOKEN_METADATA_PROGRAM_ID.toBytes(), collection.toBytes(), Buffer.from("edition")];
    const [edition,] = anchor.web3.PublicKey.findProgramAddressSync(editionSeeds, TOKEN_METADATA_PROGRAM_ID);
    const tokenAccount = await getAssociatedTokenAddress(collection, maker, true);

    // console.log(`maker ${maker.toBase58()}`)
    // console.log(`collection ${collection.toBase58()}`)
    // console.log(`metadata ${metadata.toBase58()}`)
    // console.log(`edition ${edition.toBase58()}`)

    await BRUSHO_NFT_MANAGER_PROGRAM.methods
        .initializeMaker({ issuingAuthority, updateAuthority, name, metadataUrl })
        .accounts({
            payer: payer.publicKey,
            maker,
            realm,
            governanceProgramId: GOV_PROGRAM_ID,
            realmAuthority: realmAuthority.publicKey,
            collection,
            metadata,
            masterEdition: edition,
            tokenAccount,
            tokenMetadataProgram: TOKEN_METADATA_PROGRAM_ID,
            associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
            tokenProgram: TOKEN_PROGRAM_ID
        })
        .signers([payer, realmAuthority])
        .rpc()

    return { realm, mint, councilMint, realmAuthority, maker, makerBump, collection, collectionBump, metadata, edition, tokenAccount }
}

describe("initialize_maker!", () => {
    it("verify_data", async () => {
        const authority = await newSigner();

        const name = "BrushO"
        const metadataUrl = "https://abcd.xyz/metadata";
        const { realm, mint, councilMint, realmAuthority, maker, makerBump, collection, collectionBump, metadata, edition, tokenAccount } =
            await createMaker(authority, authority.publicKey, authority.publicKey, name, metadataUrl);


        // Verify maker data
        const makerData = await BRUSHO_NFT_MANAGER_PROGRAM.account.maker.fetch(maker);
        assert.equal(realm.toBase58(), makerData.realm.toBase58())
        assert.equal(realmAuthority.publicKey.toBase58(), makerData.realmAuthority.toBase58())
        assert.equal(collection.toBase58(), makerData.collection.toBase58())
        assert.equal(anchor.web3.PublicKey.default.toBase58(), makerData.merkleTree.toBase58())
        assert.equal(authority.publicKey.toBase58(), makerData.updateAuthority.toBase58())
        assert.equal(authority.publicKey.toBase58(), makerData.issuingAuthority.toBase58())
        assert.equal(name, makerData.name)
        assert.equal(true, makerData.isActive)
        assert.equal(makerBump, makerData.bump)
        assert.equal(collectionBump, makerData.collectionBump)

        // Verify collection
        const collectionMint = await getMint(collection)
        assert.equal(0, collectionMint.decimals)
        assert.equal(edition.toBase58(), collectionMint.freezeAuthority.toBase58())
        assert.equal(edition.toBase58(), collectionMint.mintAuthority.toBase58())
        assert.equal("1", collectionMint.supply.toString())

        // Verify token account of collection
        const tokenAccountData = await getTokenAccount(tokenAccount);
        assert.equal(collection.toBase58(), tokenAccountData.mint.toBase58());
        assert.equal(maker.toBase58(), tokenAccountData.owner.toBase58());
        assert.equal("1", tokenAccountData.amount.toString());
    });

});

File: programs/brusho-nft-manager/src/instructions/initialize_maker.rs
use crate::error::BnmError;
use crate::CreateMetadataAccountsV3;
use crate::{
    create_master_edition_v3, create_metadata_accounts_v3, state::*, CreateMasterEditionV3,
    Metadata,
};
use anchor_lang::prelude::*;
use anchor_spl::{
    associated_token::AssociatedToken,
    token::{self, Mint, MintTo, Token, TokenAccount},
};
use mpl_token_metadata::types::{CollectionDetails, DataV2};
use spl_governance::state::realm::get_realm_data;

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Default)]
pub struct InitializeMakerArgs {
    pub update_authority: Pubkey,
    pub issuing_authority: Pubkey,
    pub name: String,
    pub metadata_url: String,
}

#[derive(Accounts)]
#[instruction(args: InitializeMakerArgs)]
pub struct InitializeMaker<'info> {
    #[account(mut)]
    pub payer: Signer<'info>,
    #[account(
        init,
        payer = payer,
        space = 8 + std::mem::size_of::<Maker>() + 1 + 17,
        seeds = ["maker".as_bytes(), realm.key().as_ref(), args.name.as_bytes()],
        bump,
    )]
    pub maker: Box<Account<'info, Maker>>,
    /// An spl-governance realm
    ///
    /// realm is validated in the instruction:
    /// - realm is owned by the governance_program_id
    /// - realm_authority is realm.authority
    /// CHECK:
    pub realm: UncheckedAccount<'info>,
    /// The program id of the spl-governance program the realm belongs to.
    /// CHECK:
    pub governance_program_id: UncheckedAccount<'info>,
    pub realm_authority: Signer<'info>,
    #[account(
        init,
        payer = payer,
        mint::decimals = 0,
        mint::authority = maker,
        mint::freeze_authority = maker,
        seeds = ["collection".as_bytes(), maker.key().as_ref()],
        bump
    )]
    pub collection: Box<Account<'info, Mint>>,
    /// CHECK: Handled by cpi
    #[account(
        mut,
        seeds = ["metadata".as_bytes(), token_metadata_program.key().as_ref(), collection.key().as_ref()],
        seeds::program = token_metadata_program.key(),
        bump,
    )]
    pub metadata: UncheckedAccount<'info>,
    /// CHECK: Handled by cpi
    #[account(
        mut,
        seeds = ["metadata".as_bytes(), token_metadata_program.key().as_ref(), collection.key().as_ref(), "edition".as_bytes()],
        seeds::program = token_metadata_program.key(),
        bump,
    )]
    pub master_edition: UncheckedAccount<'info>,
    #[account(
        init,
        payer = payer,
        associated_token::mint = collection,
        associated_token::authority = maker,
    )]
    pub token_account: Box<Account<'info, TokenAccount>>,

    pub token_metadata_program: Program<'info, Metadata>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, Token>,
    pub rent: Sysvar<'info, Rent>,
}

impl<'info> InitializeMaker<'info> {
    fn mint_ctx(&self) -> CpiContext<'_, '_, '_, 'info, MintTo<'info>> {
        let cpi_accounts = MintTo {
            mint: self.collection.to_account_info(),
            to: self.token_account.to_account_info(),
            authority: self.maker.to_account_info(),
        };
        CpiContext::new(self.token_program.to_account_info(), cpi_accounts)
    }
}

pub fn initialize_maker(ctx: Context<InitializeMaker>, args: InitializeMakerArgs) -> Result<()> {
    require!(args.name.len() <= 17, BnmError::InvalidMakerNameLength);
    require!(
        args.metadata_url.len() <= 200,
        BnmError::InvaliMetadataUrlLength
    );

    // Verify that "realm_authority" is the expected authority on "realm"
    let realm = get_realm_data(
        &ctx.accounts.governance_program_id.key(),
        &ctx.accounts.realm,
    )?;
    require_keys_eq!(
        realm.authority.unwrap(),
        ctx.accounts.realm_authority.key(),
        BnmError::InvalidRealmAuthority
    );

    let maker = Maker {
        realm: ctx.accounts.realm.key(),
        realm_authority: ctx.accounts.realm_authority.key(),
        collection: ctx.accounts.collection.key(),
        merkle_tree: Pubkey::default(),
        issuing_authority: args.issuing_authority,
        update_authority: args.update_authority,
        name: args.name.clone(),
        is_active: true,
        bump: ctx.bumps.maker,
        collection_bump: ctx.bumps.collection,
        reserved1: [0; 5],
        reserved2: [0; 8],
    };
    let signer_seeds: &[&[&[u8]]] = &[maker_seeds!(&maker)];

    token::mint_to(ctx.accounts.mint_ctx().with_signer(signer_seeds), 1)?;

    create_metadata_accounts_v3(
        CpiContext::new_with_signer(
            ctx.accounts
                .token_metadata_program
                .to_account_info()
                .clone(),
            CreateMetadataAccountsV3 {
                metadata: ctx.accounts.metadata.to_account_info().clone(),
                mint: ctx.accounts.collection.to_account_info().clone(),
                mint_authority: ctx.accounts.maker.to_account_info().clone(),
                payer: ctx.accounts.payer.to_account_info().clone(),
                update_authority: ctx.accounts.maker.to_account_info().clone(),
                system_program: ctx.accounts.system_program.to_account_info().clone(),
                token_metadata_program: ctx.accounts.token_metadata_program.clone(),
            },
            signer_seeds,
        ),
        DataV2 {
            name: args.name.clone(),
            symbol: "MAKER".to_string(),
            uri: args.metadata_url,
            seller_fee_basis_points: 0,
            creators: None,
            collection: None,
            uses: None,
        },
        true,
        Some(CollectionDetails::V1 { size: 0 }),
    )?;

    create_master_edition_v3(
        CpiContext::new_with_signer(
            ctx.accounts
                .token_metadata_program
                .to_account_info()
                .clone(),
            CreateMasterEditionV3 {
                edition: ctx.accounts.master_edition.to_account_info().clone(),
                mint: ctx.accounts.collection.to_account_info().clone(),
                update_authority: ctx.accounts.maker.to_account_info().clone(),
                mint_authority: ctx.accounts.maker.to_account_info().clone(),
                metadata: ctx.accounts.metadata.to_account_info().clone(),
                payer: ctx.accounts.payer.to_account_info().clone(),
                token_program: ctx.accounts.token_program.to_account_info().clone(),
                system_program: ctx.accounts.system_program.to_account_info().clone(),
                token_metadata_program: ctx.accounts.token_metadata_program.clone(),
            },
            signer_seeds,
        ),
        Some(0),
    )?;

    ctx.accounts.maker.set_inner(maker);

    Ok(())
}
